### PR TITLE
Add Blazor support to an ASP.NET Core app

### LINKED ISSUE #30471: Document how to add Auto component rendering to a Blazor Server app in .NET 8
Some Blazor Server users may want to add WebAssembly based components to their existing apps after upgrading to .NET 8. We should document how to add the client project and how to setup components to use the WebAssembly or Auto interactive render modes.

### PR DESCRIPTION
**Pull request by guardrex**
Fixes #30471

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [aspnetcore/blazor/components/integration.md](https://github.com/dotnet/AspNetCore.Docs/blob/2f363e192d4a6347e3c8b0f1809bb60f03af33d9/aspnetcore/blazor/components/integration.md) | [Integrate ASP.NET Core Razor components into ASP.NET Core apps](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/components/integration?branch=pr-en-us-30791) |

### CODE REVIEW

**Comment by guardrex** [comment]
... and let me know for the 2nd section if you want to include .......

```razor
@using static Microsoft.AspNetCore.Components.Web.RenderMode
```

.... and change the `Counter` component over to ...

```razor
@rendermode InteractiveServer
```

**Comment by danroth27** [comment]
> @danroth27 .... I take it you mean in the first two scenarios _**without a donor**_ that we'll show _**unstyled**_ components? This goes down a rabbit hole showing a lot of code without it if we show Blazor styling ... and the layout ... and the nav menu. I'll start the write-up without it (no donor, no styling, no layout, no nav menu).

Yeah, unstyled. Just the minimum required. The existing app will likely already have its own layout and styles.

File: aspnetcore/blazor/components/integration.md [hunk 1]
```diff
@@ -22,24 +22,171 @@ Razor components can be integrated into Razor Pages, MVC, and other types of ASP
...
+
+```csharp
+app.MapRazorComponents<App>();
+```
+
+When the app is run, the `Welcome` component is accessed at the `/welcome` endpoint.
+
+### Enable interactive server rendering
+
+Follow the guidance in the [Enable interactive server rendering](#enable-interactive-server-rendering) section.
```
Comment (danroth27) [inline aspnetcore/blazor/components/integration.md:137]: I think you meant to link to the prior section?

```suggestion
Follow the guidance in the [Add static server Razor component rendering](#add-static-server-razor-component-rendering) section.
```

File: aspnetcore/blazor/components/integration.md [hunk 2]
```diff
@@ -22,24 +22,171 @@ Razor components can be integrated into Razor Pages, MVC, and other types of ASP
...
+```razor
+<!DOCTYPE html>
+<html lang="en">
+
+<head>
+    <meta charset="utf-8" />
+    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+    <base href="/" />
+    <link rel="stylesheet" href="{APP NAMESPACE}.styles.css" />
+    <link rel="icon" type="image/png" href="favicon.png" />
```
Comment (danroth27) [inline aspnetcore/blazor/components/integration.md:80]: The app might not have this specific favicon.png file.

```suggestion
```

File: aspnetcore/blazor/components/integration.md [hunk 3]
```diff
@@ -22,24 +22,171 @@ Razor components can be integrated into Razor Pages, MVC, and other types of ASP
...
+@using {APP NAMESPACE}
+@using {APP NAMESPACE}.Components
+```
+
+`Components/Routes.razor`:
+
+```razor
+<Router AppAssembly="@typeof(Program).Assembly">
+    <Found Context="routeData">
+        <RouteView RouteData="@routeData" />
```
Comment (danroth27) [inline aspnetcore/blazor/components/integration.md:63]: Can we add a note about how you might want to add a default layout here that matches the rest of your app and link to the related docs on Blazor layouts?

File: aspnetcore/blazor/components/integration.md [hunk 4]
```diff
@@ -22,24 +22,171 @@ Razor components can be integrated into Razor Pages, MVC, and other types of ASP
...
+
+`Components/_Imports.razor`:
+
+```razor
+@using System.Net.Http
+@using System.Net.Http.Json
+@using Microsoft.AspNetCore.Components.Forms
+@using Microsoft.AspNetCore.Components.Routing
+@using Microsoft.AspNetCore.Components.Web
+@using Microsoft.AspNetCore.Components.Web.Virtualization
```
Comment (danroth27) [inline aspnetcore/blazor/components/integration.md:52]:
```suggestion
@using Microsoft.AspNetCore.Components.Web
@using static Microsoft.AspNetCore.Components.Web.RenderMode
@using Microsoft.AspNetCore.Components.Web.Virtualization
```

File: aspnetcore/blazor/components/integration.md [hunk 5]
```diff
@@ -22,24 +22,171 @@ Razor components can be integrated into Razor Pages, MVC, and other types of ASP
...
+    .AddInteractiveServerRenderMode();
+```
+
+Add a `Counter` component to the app with the interactive server render mode.
+
+`Components/Pages/Counter.razor`:
+
+```razor
+@page "/counter"
+@attribute [RenderModeInteractiveServer]
```
Comment (danroth27) [inline aspnetcore/blazor/components/integration.md:161]:
```suggestion
@rendermode InteractiveServer
```

**Comment by guardrex** [comment]
Hold a sec ... I'm going to merge this and add the third section back on a new PR.

I'll do more work to the first two sections on Monday.

I can get your feedback on the third section today.

*Stand-by for a sec .......................................* 🏃

**Comment by guardrex** [comment]
*Oooooops!* ... forgot to knock out the link. 🙈

File: aspnetcore/blazor/components/integration.md [hunk 6]
```diff
@@ -22,24 +22,171 @@ Razor components can be integrated into Razor Pages, MVC, and other types of ASP
...
+    private void IncrementCount()
+    {
+        currentCount++;
+    }
+}
+```
+
+When the app is run, the `Counter` component is accessed at the `/counter` endpoint.
+
+### Enable interactive Auto or WebAssembly rendering
```
Comment (danroth27) [inline aspnetcore/blazor/components/integration.md]: I'd also like to simplify this section to make it more minimalistic:
- Clarify at the beginning that the user should first enable support for static server rendering if they don't have that already, like we do in the section on enabling interactive server rendering.
- When creating the donor project use the `--empty` option. We're just trying to enable the project structure, not include the sample pages and content.
- Remove step 2 to copy over Bootstrap and the styles. The existing app might not be using Bootstrap.
- Remove step 4. The generated project without sample pages won't have a Components folder. We'll need to add a step to add a component to the client project to demonstrate that it's working.
- Simplify step 10 given that the user has done a bunch of the described changes already as part of enabling static server rendering. Key things are to enable interactive WebAssembly support and to add the client project as an additional assembly
- Step 11 is huge! 😮 I'd like to shrink it way down to just adding another *Pages/Counter2.razor* component to the client project with `@rendermode InteractiveWebAssembly` or `InteractiveAuto`.
-

**Review by danroth27 — CHANGES_REQUESTED** [review]
This is looking good! :+1: Most of my feedback is on slimming down the interactive webassembly section.

File: aspnetcore/blazor/components/integration.md (hunk 6 above)
Comment (guardrex) [reply aspnetcore/blazor/components/integration.md]: Ok ... I'll put it on the new PR tho. That way, all lines go in clean for feedback.

**Pull request merged**